Function: Automatic warning if the difficulty level is incorrect

Gogadon opened the pull request:
Hello!

This pull request adds a new, major quality-of-life feature: an automatic warning that appears when a player enters an instance with the wrong difficulty set.

**Feature Purpose:**
This feature prevents players from accidentally clearing an entire raid on the wrong difficulty, thereby losing their weekly chance at a mount. If the current difficulty does not match the requirements for a mount in the active step, a large, centered warning appears on the screen.

**Technical Implementation & Changes:**

1.  **`MRP.lua`:**
    * The `StepWatcher` now listens for all relevant zone change events (`ZONE_CHANGED_INDOORS`, `ZONE_CHANGED_NEW_AREA`, etc.) to reliably trigger the check.
    * Upon entering an instance, new logic compares the currently set difficulty with the requirements parsed from the route's notes.
    * The `MRP_MatchesDifficulty` function has been refactored to use the correct difficulty IDs and be more robust.
    * You can also view the list of difficulty level IDs here: https://wowpedia.fandom.com/wiki/DifficultyID

2.  **`MRP_UI.lua`:**
    * A new, large warning frame (`MRPDifficultyWarningFrame`) has been added to the end of the file.
    * This frame is hidden by default and is only shown when triggered by the logic in `MRP.lua`.
    * The text inside the warning frame is centered for better readability.

3.  **`MRP_Locale.enUS.lua` & `MRP_Locale.deDE.lua`:**
    * The new strings for the warning (`Wrong Difficulty` and `Please switch to '%s'`) have been added to the localization files to fully support this feature.

This feature should make the addon even more helpful for many players. Thank you for your work!

![{57B2708E-0401-4794-9784-4415B12266C7}](https://github.com/user-attachments/assets/f49a003d-e970-4d20-8205-c47648b35ae9)

Gogadon committed:
Function: Automatic warning if the difficulty level is incorrect

Hello!

This pull request adds a new, major quality-of-life feature: an automatic warning that appears when a player enters an instance with the wrong difficulty set.

**Feature Purpose:**
This feature prevents players from accidentally clearing an entire raid on the wrong difficulty, thereby losing their weekly chance at a mount. If the current difficulty does not match the requirements for a mount in the active step, a large, centered warning appears on the screen.

**Technical Implementation & Changes:**

1.  **`MRP.lua`:**
    * The `StepWatcher` now listens for all relevant zone change events (`ZONE_CHANGED_INDOORS`, `ZONE_CHANGED_NEW_AREA`, etc.) to reliably trigger the check.
    * Upon entering an instance, new logic compares the currently set difficulty with the requirements parsed from the route's notes.
    * The `MRP_MatchesDifficulty` function has been refactored to use the correct difficulty IDs and be more robust.

2.  **`MRP_UI.lua`:**
    * A new, large warning frame (`MRPDifficultyWarningFrame`) has been added to the end of the file.
    * This frame is hidden by default and is only shown when triggered by the logic in `MRP.lua`.
    * The text inside the warning frame is centered for better readability.

3.  **`MRP_Locale.enUS.lua` & `MRP_Locale.deDE.lua`:**
    * The new strings for the warning (`Wrong Difficulty` and `Please switch to '%s'`) have been added to the localization files to fully support this feature.

This feature should make the addon even more helpful for many players. Thank you for your work!

Gogadon committed:
Function: Automatic warning if the difficulty level is incorrect

Gogadon committed:
Function: Automatic warning if the difficulty level is incorrect

Gogadon committed:
Function: Automatic warning if the difficulty level is incorrect

Gogadon commented:
Okay, please wait before you take over! It always seems to output the message. I just went into a dungeon on Heroic, and I also get the message that I should switch to 25 Heroic!
I would check this again tomorrow! Sorry

Gogadon commented:
Okay, no, I'm stupid. It was still set to Icecrown Citadel in the addon!

Deathwing commented on MRP.lua:
```diff
@@ -363,16 +381,68 @@ function MRP_CheckCurrentStepComplete(force)
+        -- Hide the warning when we are not (or no longer) in an instance.
+        if not IsInInstance() then
+            MRP.UI:HideDifficultyWarning()
+            return -- End the function here, as further checks are pointless.
```
This return will result in no auto-skip checks for zon change events, therefore i.e. entering Uldum will not update to the next step correctly. 

Move your code into a new function, and call it from OnEvent, so you can do your return stuff as usual.

Gogadon replied:
Okay, I've refactored the code as you suggested. The logic is now in its own function to prevent blocking other events. Thanks for the feedback!

Deathwing reviewed (changes requested):
Just a small issue

Gogadon committed:
Refactor difficulty check into separate function

Okay, I've refactored the code as you suggested. The logic is now in its own function to prevent blocking other events. Thanks for the feedback!
